engine/runner/src/main/java/org/enso/runner/Main.java

```diff
@@ -0,0 +1,1413 @@
+    }
+  }
+
+  private void printFrame(StackFrame frame, File relativeTo) {
+    var langId = frame.isHostFrame() ? "java" : frame.getLanguage().getId();
+
+    String fmtFrame;
+    if (LanguageInfo.ID.equals(langId)) {
+      var fName = frame.getRootName();
+
+      var src = "Internal";
+      var sourceLoc = frame.getSourceLocation();
+      if (sourceLoc != null) {
+        var path = sourceLoc.getSource().getPath();
+        var ident = sourceLoc.getSource().getName();
+        if (path != null) {
+          if (relativeTo != null) {
+            var absRoot = relativeTo.getAbsoluteFile();
+            if (path.startsWith(absRoot.getAbsolutePath())) {
+              var rootDir = absRoot.isDirectory() ? absRoot : absRoot.getParentFile();
+              ident = rootDir.toPath().relativize(new File(path).toPath()).toString();
+            }
+          }
+        }
+
+        var loc = sourceLoc.getStartLine() + "-" + sourceLoc.getEndLine();
+        var line = sourceLoc.getStartLine();
+        if (line == sourceLoc.getEndLine()) {
+          var start = sourceLoc.getStartColumn();
+          var end = sourceLoc.getEndColumn();
+          loc = line + ":" + start + "-" + end;
+        }
+        src = ident + ":" + loc;
+      }
+      fmtFrame = fName + "(" + src + ")";
+    } else {
+      fmtFrame = frame.toString();
+    }
+    println("        at <" + langId + "> " + fmtFrame);
+  }
+
+  private void printPolyglotException(PolyglotException exception, scala.Option<File> relativeTo) {
+    var fullStackReversed = new LinkedList<StackFrame>();
+    for (var e : exception.getPolyglotStackTrace()) {
+      fullStackReversed.addFirst(e);
+    }
+
+    var dropInitJava =
+        fullStackReversed.stream()
+            .dropWhile(f -> !LanguageInfo.ID.equals(f.getLanguage().getId()))
+            .toList();
+    Collections.reverse(fullStackReversed);
+    var msg = HostEnsoUtils.findExceptionMessage(exception);
+    println("Execution finished with an error: " + msg);
+
+    var r = relativeTo.getOrElse(() -> (File) null);
+    if (exception.isSyntaxError()) {
+      // no stack
+    } else if (dropInitJava.isEmpty()) {
+      for (var f : exception.getPolyglotStackTrace()) {
+        printFrame(f, r);
+      }
+    } else {
+      for (var f : dropInitJava) {
+        printFrame(f, r);
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static final <T> scala.collection.immutable.List<T> nil() {
+    return (scala.collection.immutable.List<T>) scala.collection.immutable.Nil$.MODULE$;
+  }
+
+  private static final <T> scala.collection.immutable.List<T> join(
+      T head, scala.collection.immutable.List<T> tail) {
+    return scala.collection.immutable.$colon$colon$.MODULE$.apply(head, tail);
+  }
+
+  @SuppressWarnings("deprecation")
+  private static <R> R from(Either<String, R> e) throws WrongOption {
+    if (e.isLeft()) {
+      throw new WrongOption(e.left().get());
+    }
+    return e.toOption().get();
```

Comment:
Something along the lines of
```suggestion
return e.fold(failure -> throw new WrongOption(failure), v->v);
```
is easier to read and avoids conversion

Reply:
No, that doesn't work: https://github.com/enso-org/enso/actions/runs/8908274272/job/24463601026?pr=9810#step:7:1235 - I'll revert that change.